Functioning Modal Legend

mjaylene opened the pull request:
Once you click the button, the legend will open.

mjaylene committed:
Created a functioning modal! However, it is lacking the actual legend content.

mjaylene committed:
Complete functioning modal legend

mjaylene committed:
Complete first draft of modal legend.

ygina commented on dist/index.html:
```diff
@@ -16,7 +16,85 @@
+    <tr>
+      <td><img src="file:///C:/Users/jayja/karl-ui/src/img/clock.png" width="40" height="40" style="float:center;margin:0px 20px" ></td>
+      <td>Time</td> 
+      <td><div class="modual-box"></div>
```
Note: spelled "module"

mjaylene replied:
I fixed the spelling of the "module-box" class.

ygina commented on dist/index.html:
```diff
@@ -16,7 +16,85 @@
+        <div id="simpleModal" class="modal">
+          <div class="modal-content">
+            <span class="closeBtn">&times;</span>
+            <table class="legend-table">
```
`<table id="legend-table">`

mjaylene replied:
I changed the selector from "class" to "id".

ygina reviewed (changes requested):
Thanks Jaylene! In general, the legend looks good visually. Now that we know what HTML is required to display the legend, let's generate the code using TypeScript instead (in particular, the `legend-table`). As an example, note how in `index.html`, the line `<form id="module-form"></form>` doesn't have any HTML elements inside it. That's because we generated the inner elements using the `renderInitialForm()` function in `module_html.ts`, which was called from `index.ts`. The HTML elements created by this function correspond to this in the UI:

![image](https://user-images.githubusercontent.com/7017404/134963262-18f4af1c-811e-43cb-8b1a-405487502288.png)

We'll want to do something similar with the legend. You'll want to create a new file `src/legend.ts` that looks something like this:

```
export module LegendHTML {
  function genSVGArrow(color: string, dashed: bool): SVGElement {
    // since we're using TypeScript, we can write helper functions like this
    // to generate SVG arrows without writing a lot of repetitive code
    // TODO: implement
  }

  export function initialize() {
    let table = document.getElementById('legend-table')
    // TODO: create table elements and appendChild to table
  }
}
```

And then you'll want to call `LegendHTML.initialize()` from inside the `initializeLegend()` function you created.

Commit:
Typscript legend almost complete. Missing a few elements. Complete with arrows.

Commit:
Complete legend using typescript.

Commit:
Complete function legend with Typescript.

mjaylene commented:
In the last series of commits, I converted the previous HTML code used to create the legend into TypeScript. 

To do so, I created a function called genSVGArrow which takes in several arrow attributes (color, dashed, arrow name, etc.) as parameters and dynamically creates the SVG elements needed to generate the arrows (line, marker, deft, etc.). I also created multiple helper functions to dynamically generate the HTML table:

- getTableElement takes in the arrow name as input and returns the arrow with its proper attributes
- getTableText takes in the arrow name as input and returns the HTML table element td which the inner text equal to the input

To continue, I created several helper functions to append each of the elements to the table:
- appendArrow takes in the arrow name as input and returns the HTML table element td with the arrow appended onto it
- appendTableElement takes in an element name and dynamically creates the SVG elements that are not arrows (i.e. boxes representing modules/sensors)
- appendIMG dynamically creates the img tag and appends it to the td element of the table

Finally, the initialize function intializes the table elements (tbody, trTop, trBottom, etc.), appends the arrows, table text, and other SVG elements, and returns the complete table.

ygina commented on src/index.ts:
```diff
@@ -89,6 +104,8 @@ function initializeGraph() {
 
 initializeNavbar()
+initializeLegend()
+LegendHTML.initialize()
```
Put this function call inside `initializeLegend()`.

mjaylene replied:
I moved LegendHTML.initialize() within the initializeLegend() function.

ygina commented on src/legend.ts:
```diff
@@ -0,0 +1,237 @@
+export module LegendHTML {
+    function genSVGArrow(color: string, dashed: boolean, double: boolean, start: string, end: string, urlName: string, markerID: string, doubleMarkerID: string): SVGElement{
+        const svgURL = "http://www.w3.org/2000/svg"
```
If you nest this constant declaration under the module `LegendHTML` instead of this function, you can also use it when you reference the URL in `appendTableElement()`.

mjaylene replied:
I nested the svgURL constant under the module LegendHTML to avoid redundant code.

ygina commented on src/legend.ts:
```diff
@@ -0,0 +1,237 @@
+        let td = document.createElement("td")
+
+        let img = document.createElement("img")
+        img.setAttribute("src", "file:///Users/mjaylene/karl-ui/src/img/clock.png")
```
This path doesn't work on my computer because I don't have a folder called `Users/mjaylene/...`. :)

mjaylene replied:
I defined a different path to the image that does not use my own personal folder.

ygina reviewed (changes requested):
1. Can you remove the two `.DS_Store` files from git tracking? This file isn't needed on other people's computers so there is no need to track it.
2. And can you include the line `.DS_Store` in the `.gitignore` file to automatically ignore the file in the future?
3. What is the purpose of the new `karl-ui` submodule?

Commit:
Fully functioning modal legend with requested changes:making use of svgURL constant, moving LegendHTML.initialize() within the initializeLegend() function, and changing the path of the clock image.

mjaylene committed:
In the last series of commits, I converted the previous HTML code used to create the legend into TypeScript.

To do so, I created a function called genSVGArrow which takes in several arrow attributes (color, dashed, arrow name, etc.) as parameters and dynamically creates the SVG elements needed to generate the arrows (line, marker, deft, etc.). I also created multiple helper functions to dynamically generate the HTML table:

getTableElement takes in the arrow name as input and returns the arrow with its proper attributes
getTableText takes in the arrow name as input and returns the HTML table element td which the inner text equal to the input
To continue, I created several helper functions to append each of the elements to the table:

appendArrow takes in the arrow name as input and returns the HTML table element td with the arrow appended onto it
appendTableElement takes in an element name and dynamically creates the SVG elements that are not arrows (i.e. boxes representing modules/sensors)
appendIMG dynamically creates the img tag and appends it to the td element of the table
Finally, the initialize function intializes the table elements (tbody, trTop, trBottom, etc.), appends the arrow, table text, and other SVG elements, and returns the complete table.

mjaylene committed:
Removed the two .DS_Store files from git tracking.

mjaylene committed:
Removed the .DS_Store from tracking and ignored it.

mjaylene committed:
Deleted the karl-ui submodule.

ygina commented on src/css/style.css:
```diff
@@ -173,18 +173,54 @@ line {
 }
 
 .modal {
```
Instead of modifying the existing `.modal` class, can you create a new class `.legend-modal` for the legend? Adding properties here will modify existing elements that are assigned this class (such as in `host_html.ts`) in unintended ways. And then make sure to assign the legend modal you created with this different class name, instead of `.modal`.

ygina commented on src/css/style.css:
```diff
@@ -173,18 +173,54 @@ line {
   margin-bottom: 5px;
 }
 
 .modal p {
```
Same here. If you change the existing CSS it can affect existing modals. If you want to increase the margin _just_ for the legend modal, you should create a new tag `.legend-modal p { <fields> }`.

ygina commented on src/css/style.css:
```diff
@@ -173,18 +173,54 @@ line {
   padding-left: 5px;
 }
+
+.modal-content{
```
To avoid confusion with non-legend modals, I would rename this `.legend-modal-content`.

ygina commented on src/legend.ts:
```diff
@@ -0,0 +1,239 @@
+        line.setAttribute("stroke-width", "8")
+        line.setAttribute("marker-end", urlName)
+
+         if (dashed) {
```
Can you follow the spacing conventions of the rest of the code, and put any `else` or `else if` statements on the same line as the closing curly brace? i.e.,

```
if (dashed) {
    ...
} else if (double) {
    ...
}
```

ygina commented on src/legend.ts:
```diff
@@ -0,0 +1,239 @@
+        let td = document.createElement("td")
+
+        let img = document.createElement("img")
+        img.setAttribute("src", "img/clock.png")
```
Hmm the image still doesn't show up for me. How do other parts of the code refer to `clock.png`?

mjaylene committed:
To avoid interfering with other modal elements in the UI, I created a new class .legend-modal. I also modified the name of the class name of the containiner that holds the legend to be legend-modal-content to specify its purpose. In doing so, I updated the css to reflect the changes in class names. The spacing in legend.ts was also fixed; I aligned everything and converted the indentation to spaces. To resolve the issue of the clock not showing up, I followed the approach used to add a clock icon in the graph_html.ts file. I created an image with the Image() function and specified the path to the image with the require() function.

mjaylene committed:
In this series of changes, I created a new class of modal for the legend specifically. I updated the css to reflect the new class, added back in the previous modal css code, and created my own legend-modal style choices. I also updated the html and typescript code to, once again, reflect the changes.

mjaylene committed:
Fixed small errors in code

ygina commented:
Looks good! :)

ygina committed:
Merge branch 'main' into main
